pkg/apis/nfd/validate/validate.go

```diff
@@ -0,0 +1,177 @@
+		nameSplit := strings.SplitN(match.Feature, ".", 2)
+		if len(nameSplit) != 2 {
+			validationErr = append(validationErr, fmt.Errorf("invalid feature name %v (not <domain>.<feature>), cannot be used for templating", match.Feature))
+		}
+	}
+
+	return validationErr
+}
+
+func Template(labelsTemplate string) []error {
+	var validationErr []error
+
+	// Validate template
+	_, err := template.New("").Option("missingkey=error").Parse(labelsTemplate)
+	if err != nil {
+		validationErr = append(validationErr, fmt.Errorf("invalid template: %w", err))
+	}
+	return validationErr
+}
+
+func Annotations(annotations map[string]string) []error {
+	var validationErr []error
+
+	for key, value := range annotations {
+		validationErr = append(validationErr, Annotation(key, value)...)
+	}
+
+	return validationErr
+}
+
+func Annotation(key, value string) []error {
+	var validationErr []error
+
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureAnnotationNs && !strings.HasSuffix(ns, nfdv1alpha1.FeatureAnnotationSubNsSuffix) {
+		// If the namespace is denied, rule is invalid
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			validationErr = append(validationErr, fmt.Errorf("namespace %v is not allowed", ns))
+		}
+	}
+	// Validate the annotation value
+	if errs := k8svalidation.IsValidLabelValue(value); len(errs) > 0 {
+		validationErr = append(validationErr, fmt.Errorf("invalid value %q: %s", value, strings.Join(errs, "; ")))
+	}
+
+	return validationErr
+}
+
+func Labels(labels map[string]string) []error {
+	var validationErr []error
+
+	for key, value := range labels {
+		validationErr = append(validationErr, Label(key, value)...)
+	}
+
+	return validationErr
+}
+
+func Label(key, value string) []error {
+	var validationErr []error
+
+	ns, _ := splitNs(key)
+	if ns != nfdv1alpha1.FeatureLabelNs && !strings.HasSuffix(ns, nfdv1alpha1.FeatureLabelSubNsSuffix) {
+		// If the namespace is denied, rule is invalid
+		if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
+			validationErr = append(validationErr, fmt.Errorf("namespace %v is not allowed", ns))
+		}
+	}
+
+	// Validate the label value
+	if errs := k8svalidation.IsValidLabelValue(value); len(errs) > 0 {
+		validationErr = append(validationErr, fmt.Errorf("invalid value %q: %s", value, strings.Join(errs, "; ")))
+	}
+
+	return validationErr
+}
+
+func Taints(taints []corev1.Taint) []error {
+	var validationErr []error
+
+	if len(taints) < 1 {
+		// nothing to do here
+		return validationErr
+	}
+
+	for _, taint := range taints {
+		ns, _ := splitNs(taint.Key)
+		// if ns is empty, it means that the rule is not allowed
+		if ns == "" {
+			validationErr = append(validationErr, fmt.Errorf("unprefixed keys are disallowed"))
+		}
+		if ns != nfdv1alpha1.TaintNs && !strings.HasSuffix(ns, nfdv1alpha1.TaintSubNsSuffix) {
+			// If the namespace is denied, rule is invalid
+			if ns == "kubernetes.io" || strings.HasSuffix(ns, ".kubernetes.io") {
```

Comment:
You are using this if statement couple of times maybe make it a  `func` e.g. `IsValidNamespace` or `IsAllowedNamespace`?

Reply:
I thought about that, but is a one liner, and if you look the variables are different. reusing with a Func will look really similar, since the func would have to have 2 entries and return a bool.